check-disk/lib/check-disk.go

```diff
@@ -0,0 +1,148 @@
+package checkdisk
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/jessevdk/go-flags"
+	"github.com/mackerelio/checkers"
+)
+
+type diskStatus struct {
+	All   uint64
+	Used  uint64
+	Free  uint64
+	Avail uint64
+}
+
+var opts struct {
+	Warning  *string `short:"w" long:"warning" value-name:"N, N%" description:"Exit with WARNING status if less than N units or N% of disk are free"`
+	Critical *string `short:"c" long:"critical" value-name:"N, N%" description:"Exit with CRITICAL status if less than N units or N% of disk are free"`
+	Path     *string `short:"p" long:"path" value-name:"PATH" description:"Mount point or block device as emitted by the mount(8) command"`
+	Units    *string `short:"u" long:"units" value-name:"STRING" description:"Choose bytes, kB, MB, GB, TB (default: MB)"`
+}
+
+const (
+	b  = float64(1)
+	kb = float64(1024) * b
+	mb = float64(1024) * kb
+	gb = float64(1024) * mb
+	tb = float64(1024) * gb
+)
+
+func getDiskUsage(path string) (*diskStatus, error) {
+	fs := syscall.Statfs_t{}
+	err := syscall.Statfs(path, &fs)
+	if err != nil {
+		return nil, err
+	}
+
+	disk := &diskStatus{}
+	disk.All = fs.Blocks * uint64(fs.Bsize)
+	disk.Free = fs.Bfree * uint64(fs.Bsize)
+	disk.Used = disk.All - disk.Free
+	disk.Avail = fs.Bavail * uint64(fs.Bsize)
+
+	return disk, nil
+}
+
+func checkStatus(threshold string, units float64, disk *diskStatus, status checkers.Status) (checkers.Status, error) {
+	avail := float64(disk.Avail) / float64(units)
+	freePct := (float64(disk.Avail) * float64(100)) / float64(disk.All)
+
+	checkSt := checkers.OK
+	if strings.HasSuffix(threshold, "%") {
+		v, err := strconv.ParseFloat(strings.TrimRight(threshold, "%"), 64)
+		if err != nil {
+			return checkers.UNKNOWN, err
+		}
+
+		if v > freePct {
+			checkSt = status
+		}
+	} else {
+		v, err := strconv.ParseFloat(threshold, 64)
+		if err != nil {
+			return checkers.UNKNOWN, err
+		}
+
+		if v > avail {
+			checkSt = status
+		}
+	}
+
+	return checkSt, nil
+}
+
+// Do the plugin
+func Do() {
+	ckr := run(os.Args[1:])
+	ckr.Name = "Disk"
+	ckr.Exit()
+}
+
+func run(args []string) *checkers.Checker {
+	_, err := flags.ParseArgs(&opts, args)
+	if err != nil {
+		os.Exit(1)
+	}
+
+	path := "/"
```

Comment:
Hmm…
nagios's check_disk checks against all partitions by default, so this behavior may be confusing.